File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = preorder.size();
        for(int i = 0; i < inorder.size(); i++){
            inorderIndex[inorder[i]] = i;
        }
        return buildTreeCore(preorder, inorder, 0, n - 1, 0, n -1);
    }

private:
    TreeNode* buildTreeCore(vector<int>& preorder, vector<int>& inorder, int preLeft, int preRight, int inLeft, int inRight){
        if(preLeft > preRight) 
            return nullptr;
        // get the index of root in inorder array.
        int inRoot = inorderIndex[preorder[preLeft]];
        // create root node.
        auto root = new TreeNode(preorder[preLeft]);
        // get the length of left subtree.
        int len = inRoot - inLeft;
        // build left and right subtree.
        root->left = buildTreeCore(preorder, inorder, 
            preLeft + 1, preLeft + len, 
            inLeft, inRoot - 1);
        root->right = buildTreeCore(preorder, inorder, 
            preLeft + len + 1, preRight, 
            inRoot + 1, inRight);
        return root;
    }

private:
    unordered_map<int, int> inorderIndex;
};

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Map<Integer, Integer> inorderIndex;

    private TreeNode buildTreeCore(int[] preorder, int[] inorder, int preLeft, int preRight, int inLeft, int inRight){
        if(preLeft > preRight) 
            return null;
        // get the index of root in inorder array.
        int inRoot = inorderIndex.getOrDefault(preorder[preLeft], -1);
        // create root node.
        var root = new TreeNode(preorder[preLeft]);
        // get the length of left subtree.
        int len = inRoot - inLeft;
        // build left and right subtree.
        root.left = buildTreeCore(preorder, inorder, 
            preLeft + 1, preLeft + len, 
            inLeft, inRoot - 1);
        root.right = buildTreeCore(preorder, inorder, 
            preLeft + len + 1, preRight, 
            inRoot + 1, inRight);
        return root;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        inorderIndex = new HashMap<Integer, Integer>();
        for(int i = 0; i < inorder.length; i++){
            inorderIndex.put(inorder[i], i);
        }
        return buildTreeCore(preorder, inorder, 0, n - 1, 0, n -1);
    }

}







